allennlp/modules/span_extractors/bidirectional_endpoint_span_extractor.py

```diff
@@ -0,0 +1,230 @@
+        span width features.
+    span_width_embedding_dim : ``int``, optional (default = None).
+        The embedding size for the span_width features.
+    bucket_widths : ``bool``, optional (default = False).
+        Whether to bucket the span widths into log-space buckets. If ``False``,
+        the raw span widths are used.
+    use_sentinels : ``bool``, optional (default = ``True``).
+        If ``True``, sentinels are used to represent exclusive span indices for the elements
+        in the first and last positions in the sequence (as the exclusive indices for these
+        elements are outside of the the sequence boundary). This is not strictly necessary,
+        as you may know that your exclusive start and end indices are always within your sequence
+        representation, such as if you have appended/prepended <START> and <END> tokens to your
+        sequence (equivelent to learning a sentinel representation in this class, but more accessable).
+    """
+    def __init__(self,
+                 input_dim: int,
+                 forward_combination: str = "y-x",
+                 backward_combination: str = "x-y",
+                 num_width_embeddings: int = None,
+                 span_width_embedding_dim: int = None,
+                 bucket_widths: bool = False,
+                 use_sentinels: bool = True) -> None:
+        super().__init__()
+        self._input_dim = input_dim
+        self._forward_combination = forward_combination
+        self._backward_combination = backward_combination
+        self._num_width_embeddings = num_width_embeddings
+        self._bucket_widths = bucket_widths
+
+        if self._input_dim % 2 != 0:
+            raise ConfigurationError("The input dimension is not divisible by 2, but the "
+                                     "BidirectionalEndpointSpanExtractor assumes the embedded representation "
+                                     "is bidirectional (and hence divisible by 2).")
+        if num_width_embeddings is not None and span_width_embedding_dim is not None:
+            self._span_width_embedding = Embedding(num_width_embeddings, span_width_embedding_dim)
+        elif not all([num_width_embeddings is None, span_width_embedding_dim is None]):
+            raise ConfigurationError("To use a span width embedding representation, you must"
+                                     "specify both num_width_buckets and span_width_embedding_dim.")
+        else:
+            self._span_width_embedding = None
+
+        self._use_sentinels = use_sentinels
+        if use_sentinels:
+            self._start_sentinel = Parameter(torch.randn([1, 1, int(input_dim / 2)]))
+            self._end_sentinel = Parameter(torch.randn([1, 1, int(input_dim / 2)]))
+
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    def get_output_dim(self) -> int:
+        unidirectional_dim = int(self._input_dim / 2)
+        forward_combined_dim = get_combined_dim(self._forward_combination,
```

Comment:
Slight preference for importing the module and using `util.get_combined_dim` instead of this - it makes the imports nicer above, and it makes it more obvious where this function is coming from as I'm reading this.  It's not a strong preference though, so if you like it better this way, keep it, and I'll try to remember to not bring it up again.

Reply:
No this is good, I was just doing this incrementally and the imports crept up on me.